scrapy/utils/reqser.py

```diff
@@ -2,12 +2,16 @@
 Helper functions for serializing (and deserializing) requests.
 """
 import six
+import re
 
 from scrapy.http import Request
 from scrapy.utils.python import to_unicode, to_native_str
 from scrapy.utils.misc import load_object
 
 
+private_name_regex = re.compile('^__[^_](.*[^_])?_?$')
```

Comment:
I'm not sure I understand the use of `^` in the middle of the pattern. IIRC `^` means the start of the string (https://docs.python.org/3/library/re.html#regular-expression-syntax). Also, `.*` without `?` is greedy so it would match everything, including trailing underscores.
What about `^___*.*?_?$`?

Reply:
This is not working for mw with the following:
```python
re.match('^__[^_](.*[^_])?_?$', '___private_stuff_')
```
I think this is difficult to do just with a regex, because you need to allow an arbitrary amount of underscores in the middle of the identifier. What about something like `'^(_*).*?(_*)$'`, which captures underscores at the beginning and end, and then you count the amount?
I'm also concerned about the performance impact of applying the regex, this would be done for each request that goes through the scheduler.

Reply:
> I'm not sure I understand the use of `^` in the middle of the pattern. IIRC `^` means the start of the string (https://docs.python.org/3/library/re.html#regular-expression-syntax). Also, `.*` without `?` is greedy so it would match everything, including trailing underscores.

`^` at the beginning of `[…]` means that it matches any character _not_ in `[…]`.

Reply:
The only issue with `'^__[^_](.*[^_])?_?$'` is that it does not allow more than 2 underscores at the beginning.

What about `'^__+.*[^_]_?$'`?

Reply:
Thanks!  I updated the regex and did something basically like that, and added a unit test to test the regex directly since it was easier doing that than creating a bunch of classes/callbacks to test various name underscore combinations.

I'll try the bench - if this is too slow maybe it could be memoized?  Function names are typically constant through the job.

Reply:
True, thanks @Gallaecio! It's explained in the same section of the docs I linked :man_facepalming:
That also explains the _greedy_ situation I mentioned.
Anyway, now I wonder if maybe using regex is too expensive? What about just `name.startswith('__') and not name.endswith('__')`?

```python
In [1]: import re 
   ...: private_name_regex = re.compile('^__.*[^_]_?$') 
   ...: names = ['__foo', 'foo_bar', '__foo_bar_', '___foobar__']                                                                                                                                                                             

In [2]: %timeit [private_name_regex.search(n) for n in names]                                                                                                                                                                                 
1.45 µs ± 40.8 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [3]: %timeit [n.startswith('__') and not n.endswith('__') for n in names]                                                                                                                                                                  
806 ns ± 1.38 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)ndswith('__') for n in names]
```

Reply:
> if this is too slow maybe it could be memoized? Function names are typically constant through the job.

I agree, creating a separate function and using `functools.lru_cache` looks like a good option here.

Reply:
Sorry, didn't get to the benchmarking but I switched to the `startswith`/`endswith` solution because it's easier to read and faster :) and all the tests pass.

Still important to memoize?

Reply:
I think it would be good to memoize, you're right that several requests will have the same callback so it's probably not necessary to check them every time.

Reply:
Not what I was originally going to bench, but benched with/without lru_cache:

**With lru_cache**
```
Test = 'Book Spider' Iterations = '10'


Mean : 65.64734686683296 Median : 65.72720576665657 Std Dev : 0.4875654825761466
```

**Without lru_cache**
```
Test = 'Book Spider' Iterations = '10'


Mean : 65.23675882544786 Median : 65.02926278845428 Std Dev : 0.6786585837183575
```

Both were using the code on this branch otherwise.

So my computer isn't exactly a clean bench environment and the samples are fairly close -- according to [this :D](http://courses.umass.edu/pubp608/lectures/l6.pdf), I think any optimization at this level is probably lost in the noise.  Also, I lru_cache'd `_find_method` which might be beyond the scope of this change - maybe optimizing function lookup could be a separate PR?

Reply:
Yeah that sounds good. Also, sorry for bringing `functools.lru_cache` into the conversation without checking its availability first, but it was added in py3.2 and we're not dropping support for py2 just yet :-/